lib/internal/encoding.js

```diff
@@ -533,48 +532,52 @@ function makeTextDecoderJS() {
 }
 
 // Mix in some shared properties.
-ObjectDefineProperties(
-  TextDecoder.prototype,
-  ObjectFromEntries(ArrayPrototypeMap(ObjectEntries(ObjectGetOwnPropertyDescriptors({
-    get encoding() {
-      validateDecoder(this);
-      return this[kEncoding];
-    },
-
-    get fatal() {
-      validateDecoder(this);
-      return (this[kFlags] & CONVERTER_FLAGS_FATAL) === CONVERTER_FLAGS_FATAL;
-    },
-
-    get ignoreBOM() {
-      validateDecoder(this);
-      return (this[kFlags] & CONVERTER_FLAGS_IGNORE_BOM) ===
+const sharedProperties = ObjectGetOwnPropertyDescriptors({
+  get encoding() {
+    validateDecoder(this);
+    return this[kEncoding];
+  },
+
+  get fatal() {
+    validateDecoder(this);
+    return (this[kFlags] & CONVERTER_FLAGS_FATAL) === CONVERTER_FLAGS_FATAL;
+  },
+
+  get ignoreBOM() {
+    validateDecoder(this);
+    return (this[kFlags] & CONVERTER_FLAGS_IGNORE_BOM) ===
               CONVERTER_FLAGS_IGNORE_BOM;
-    },
+  },
 
-    [inspect](depth, opts) {
-      validateDecoder(this);
-      if (typeof depth === 'number' && depth < 0)
-        return this;
-      const constructor = getConstructorOf(this) || TextDecoder;
-      const obj = ObjectCreate({ constructor });
-      obj.encoding = this.encoding;
-      obj.fatal = this.fatal;
-      obj.ignoreBOM = this.ignoreBOM;
-      if (opts.showHidden) {
-        obj[kFlags] = this[kFlags];
-        obj[kHandle] = this[kHandle];
-      }
-      // Lazy to avoid circular dependency
-      const { inspect } = require('internal/util/inspect');
-      return `${constructor.name} ${inspect(obj)}`;
+  [inspect](depth, opts) {
+    validateDecoder(this);
+    if (typeof depth === 'number' && depth < 0)
+      return this;
+    const constructor = getConstructorOf(this) || TextDecoder;
+    const obj = ObjectCreate({ constructor });
+    obj.encoding = this.encoding;
+    obj.fatal = this.fatal;
+    obj.ignoreBOM = this.ignoreBOM;
+    if (opts.showHidden) {
+      obj[kFlags] = this[kFlags];
+      obj[kHandle] = this[kHandle];
     }
-  })), ({ 0: key, 1: desc }) => [key, { __proto__: null, ...desc }]))
-);
+    // Lazy to avoid circular dependency
+    const { inspect } = require('internal/util/inspect');
+    return `${constructor.name} ${inspect(obj)}`;
+  }
+});
+const propertiesValues = ObjectValues(sharedProperties);
+for (let i = 0; i < propertiesValues.length; i++) {
+  // We want to use null-prototype objects to not rely on globally mutable
+  // %Object.prototype%.
+  ObjectSetPrototypeOf(propertiesValues[i], null);
+}
+sharedProperties[inspect].enumerable = false;
```

Comment:
Why not mutate it instead? If mutating is not an option, I think I'd prefer `ObjectDefineProperty(TextDecoder.prototype, inspect, { __proto__: null, enumerable: false });`

Reply:
mutation is something that both slows things down (by changing the object's shape) and is also conceptually distasteful, so in general it's nicer to create objects wholesale rather than ever altering one.

go with whatever you prefer, obviously

Reply:
Unless someone else wants to take a side, I think I'm going to keep it as is. I'm fine with whichever solution, but mutating `sharedProperties` seems the most natural to me in this instance.